src/Microsoft.Health.Fhir.Core/Features/Validation/NarrativeValidator.cs

```diff
@@ -31,50 +31,51 @@ public override IEnumerable<ValidationFailure> Validate(PropertyValidatorContext
 
             if (context.PropertyValue is ResourceElement resourceElement)
             {
-                var poco = resourceElement.ToPoco();
-
-                if (poco is DomainResource resource)
+                if (resourceElement.IsDomainResource)
                 {
-                    foreach (ValidationFailure validationFailure in ValidateResource(resource))
+                    foreach (ValidationFailure validationFailure in ValidateResource(resourceElement))
                     {
                         yield return validationFailure;
                     }
                 }
-
-                if (poco is Bundle bundle)
+                else if (resourceElement.InstanceType.Equals(KnownResourceTypes.Bundle, System.StringComparison.OrdinalIgnoreCase))
                 {
-                    var domainResources = bundle.Entry.Select(x => x.Resource).OfType<DomainResource>();
-
-                    foreach (ValidationFailure validationFailure in domainResources.SelectMany(ValidateResource))
+                    var bundleEntries = resourceElement.Instance.Select(KnownFhirPaths.BundleEntries);
+                    if (bundleEntries != null)
                     {
-                        yield return validationFailure;
+                        var domainResources = bundleEntries.Select(e => e.ToResourceElement()).Where(r => r.IsDomainResource);
+                        foreach (ValidationFailure validationFailure in domainResources.SelectMany(ValidateResource))
+                        {
+                            yield return validationFailure;
+                        }
```

Comment:
I still think we should combine this:

```
var domainResources = bundleEntries.Select(e => e.ToResourceElement()).Where(r => r.IsDomainResource);
foreach (ValidationFailure validationFailure in domainResources.SelectMany(ValidateResource))
{
    yield return validationFailure;
}
```
to just be:

```
foreach (ValidationFailure validationFailure in bundleEntries.SelectMany(ValidateResource))
{
    yield return validationFailure;
}
```

Given that this is just private it should be ok to use ITypedElement since we just want to select out the text
```
private IEnumerable<ValidationFailure> ValidateResource(ITypedElement domainResource)
```

Previously we cared if the type was DomainResource since it is the base for `Text.Div`. FhirPath doesn't care about the type, it only seeks the path in the data.

Thoughts?